lib/history/memory/event.go

```diff
@@ -187,11 +187,12 @@ type nodeAdded struct {
 
 func (r *nodeAdded) Insert(ctx context.Context, execer history.Execer) error {
 	const insertStmt = "timestamp,type,node"
-	e := r.GetNodeAdded()
-	if e == nil {
-		return trace.BadParameter("expected NodeAdded, got nil")
+	data, ok := r.GetData().(*pb.TimelineEvent_NodeAdded)
```

Comment:
Not sure why you would keep the boxed `Data` inside specific event types:
```go
// newDataInserter returns the event as a history.DataInserter.
func newDataInserter(event *pb.TimelineEvent) (row history.DataInserter, err error) {
	switch t := event.GetData().(type) {
	case *pb.TimelineEvent_ClusterDegraded:
		return &clusterDegraded{ts: event.Timestamp}, nil
	case *pb.TimelineEvent_ClusterHealthy:
		return &clusterHealthy{ts: event.Timestamp}, nil
	case *pb.TimelineEvent_NodeAdded:
		return &nodeAdded{ts: event.Timestamp, data: t.NodeAdded}, nil
	case *pb.TimelineEvent_NodeRemoved:
		return &nodeRemoved{ts: event.Timestamp, data: t.NodeRemoved}, nil
	case *pb.TimelineEvent_NodeHealthy:
		return &nodeHealthy{ts: event.Timestamp, data: t.NodeHealthy}, nil
	case *pb.TimelineEvent_NodeDegraded:
		return &nodeDegraded{ts: event.Timestamp, data: t.NodeDegraded}, nil
	case *pb.TimelineEvent_ProbeSucceeded:
		return &probeSucceeded{ts: event.Timestamp, data: t.ProbeSucceeded}, nil
	case *pb.TimelineEvent_ProbeFailed:
		return &probeFailed{ts: event.Timestamp, data: t.ProbeFailed}, nil
	case *pb.TimelineEvent_LeaderElected:
		return &leaderElected{ts: event.Timestamp, data: t.LeaderElected}, nil
	default:
		return row, trace.BadParameter("unknown event type %T", t)
	}
}
// ...
// clusterDegraded represents a cluster degraded event.
//
// Implements history.DataInserter.
type clusterDegraded struct {
	ts *pb.Timestamp
}
// ...
// nodeAdded represents a node added event.
//
// Implements history.DataInserter.
type nodeAdded struct {
	ts   *pb.Timestamp
	data *pb.NodeAdded
}
// ...
func (r *nodeAdded) Insert(ctx context.Context, execer history.Execer) error {
	const insertStmt = "timestamp,type,node"
	return trace.Wrap(execer.Exec(ctx, insertStmt, r.ts.ToTime(), history.NodeAdded, r.data.Node))
}
```
Likewise for the rest of the event types.

Reply:
Ah this makes more sense.
Made changes in 8d81ddf